### PR TITLE
[SYCL] Support for arrays as kernel parameters

### PR DESCRIPTION
**Pull request by rdeodhar**
Signed-off-by: rdeodhar <rajiv.deodhar@intel.com>

### CODE REVIEW

**Comment by erichkeane** [comment]
@romanovvlad  As discussed previously via email, I want @rolandschulz to review the design/interface/IR output, so we're waiting on the submitter to provide the design documentation for this.  Additionally, @elizabethandrews  and @Fznamznon know this section of code better than me, so I'll delegate review to them.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 1]
```diff
@@ -986,43 +1107,52 @@ static bool buildArgTys(ASTContext &Context, CXXRecordDecl *KernelObj,
...
+              /*Id=*/nullptr, ArgTy,
+              Context.getTrivialTypeSourceInfo(ArgTy, SourceLocation()),
+              /*BW=*/nullptr, /*Mutable=*/false, /*InitStyle=*/ICIS_NoInit);
+          Field->setAccess(AS_public);
+          NewClass->addDecl(Field);
+          NewClass->completeDefinition();
+        };
+        RecordDecl *NewClass;
+        FieldDecl *Field;
+        wrapAnArray(ArgTy, NewClass, Field);
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp]: Did I get it right that to pass array of something, we wrap it by some struct (which seems to be generated by the compiler)? i.e. resulting kernel for each array field of kernel object has some `wrapped_array` struct parameter?

File: clang/lib/Sema/SemaSYCL.cpp [hunk 2]
```diff
@@ -1139,6 +1295,31 @@ static void populateIntHeader(SYCLIntegrationHeader &H, const StringRef Name,
...
+        const ASTRecordLayout &AccessorLayout =
+            AccessorCtx.getASTRecordLayout(Accessor);
+        auto AccessorSize = AccessorCtx.toBits(AccessorLayout.getSize()) / 8;
+        // Start with offset of array, i.e., offset of first element of array
+        uint64_t OffsetInArray = Offset;
+        // Treat each element of accessor array as a separate item
+        // Create a parameter descriptor for it
+        for (size_t I = 0; I < NumElements; ++I) {
+          populateHeaderForAccessor(ET, OffsetInArray);
+          OffsetInArray += AccessorSize;
```
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp]: After lines 1231-1253 it's déjà vu again

File: clang/test/SemaSYCL/array-kernel-param-run.cpp [hunk 3]
```diff
@@ -0,0 +1,259 @@
+// This test checks array kernel parameters
```
[the remark this answers is not in the document]
Comment (rdeodhar) [reply clang/test/SemaSYCL/array-kernel-param-run.cpp]: I will write a design document to explain the overall approach to supporting array arguments of kernels. I assume attaching documents here is accepted practice?
I will also address comments posted here in my next attempt.

**Comment by elizabethandrews** [comment]
From what I undestand you've handled 3 types of kernel parameters here - 

1. Accessor arrays
2. Arrays of other types
3. Class type with an accessor array as a member. 

For 1. and 3.,  you added each element as a separate parameter in generated OpenCL Kernel and called init functions accordingly.

For 2. You wrapped the array inside a new class 'wrapped _array' and added this as the parameter. 

Is this a gist of what you did? Am I missing anything?  It would be helpful having this information in the commit message.

**Comment by elizabethandrews** [comment]
Please add a Sema test showing AST generated.

File: clang/test/SemaSYCL/array-kernel-param-run.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/test/SemaSYCL/array-kernel-param-run.cpp]: Sure, we appreciate design docs. You can put your doc somewhere in this folder https://github.com/intel/llvm/tree/sycl/sycl/doc (or extend some existing doc here).

[the remark this answers is not in the document]
Comment (rdeodhar) [reply clang/test/SemaSYCL/array-kernel-param-run.cpp]: OK

File: clang/test/SemaSYCL/array-kernel-param-neg.cpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -fcxx-exceptions -verify -fsyntax-only %s
+
+// This test checks if compiler reports compilation error on an attempt to pass
+// a struct with non-trivially copyable type as SYCL kernel parameter.
```
[the remark this answers is not in the document]
Comment (rdeodhar) [reply clang/test/SemaSYCL/array-kernel-param-neg.cpp]: Yes, comment changed.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 5]
```diff
@@ -1095,6 +1225,32 @@ static void populateIntHeader(SYCLIntegrationHeader &H, const StringRef Name,
               populateHeaderForWrappedAccessors(FldType,
                                                 Offset + OffsetInWrapper);
             }
+          } else if (FldType->isConstantArrayType()) {
+            auto CAT = cast<ConstantArrayType>(FldType);
+            auto ET = CAT->getElementType();
```
[the remark this answers is not in the document]
Comment (rdeodhar) [reply clang/lib/Sema/SemaSYCL.cpp]: OK

File: clang/lib/Sema/SemaSYCL.cpp [hunk 6]
```diff
@@ -853,15 +884,63 @@ static CompoundStmt *CreateOpenCLKernelBody(Sema &S,
...
+          // Call its __init function
+          for (size_t I = 0; I < NumElements; ++I) {
+            ExprResult IndexExpr = S.ActOnIntegerConstant(Loc, I);
+            auto MA = S.CreateBuiltinArraySubscriptExpr(Lhs, Loc,
+                                                        IndexExpr.get(), Loc);
+            getExprForSpecialSYCLObj(ET, nullptr, MA.get(),
+                                     ET->getAsCXXRecordDecl(), nullptr,
+                                     InitMethodName, BodyStmts);
+            KernelFuncParam++;
+          }
```
[the remark this answers is not in the document]
Comment (rdeodhar) [reply clang/lib/Sema/SemaSYCL.cpp]: These two blocks of code are superficially similar. The check for Accessor type and then iterating over the array elements is common. However, the actions in the two cases are very different. It is not practical to make a callable function out of these two blocks of code.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 7]
```diff
@@ -853,15 +884,63 @@ static CompoundStmt *CreateOpenCLKernelBody(Sema &S,
...
             // Will put it later to the end of function body.
-            getExprForSpecialSYCLObj(FieldType, Field, CRD, KernelObjCloneRef,
-                                     FinalizeMethodName, FinalizeStmts);
+            getExprForSpecialSYCLObj(FieldType, Field, nullptr, CRD,
+                                     KernelObjCloneRef, FinalizeMethodName,
+                                     FinalizeStmts);
           }
         }
+      } else if (FieldType->isConstantArrayType()) {
+        auto Loc = Field->getLocation();
```
[the remark this answers is not in the document]
Comment (rdeodhar) [reply clang/lib/Sema/SemaSYCL.cpp]: OK. This was done based on a previous reviewers suggestion. I have reverted to empty source location to conform to the approach in other cases in this function.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 1 above)
Comment (rdeodhar) [reply clang/lib/Sema/SemaSYCL.cpp]: Yes, a design document will be posted.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 2 above)
Comment (rdeodhar) [reply clang/lib/Sema/SemaSYCL.cpp]: In this case, there was somewhat more similarity in the actions done in the two cases, and this code has been factored out into a separate function, called twice.

**Comment by rdeodhar** [comment]
Yes, that is the gist of the changes. I have uploaded a document that describes this change.

**Comment by erichkeane** [comment]
> From what I undestand you've handled 3 types of kernel parameters here -
> 
>     1. Accessor arrays
> 
>     2. Arrays of other types
> 
>     3. Class type with an accessor array as a member.
> 
> 
> For 1. and 3., you added each element as a separate parameter in generated OpenCL Kernel and called init functions accordingly.
> 
> For 2. You wrapped the array inside a new class 'wrapped _array' and added this as the parameter.
> 
> Is this a gist of what you did? Am I missing anything? It would be helpful having this information in the commit message.
Hmm.. I cannot help but wonder what happens when multiple arrays are passed?  Are they in separate types?  Are those types necessary?  I think LLVM-IR supports array-level parameters (despite C/C++ not ), so I wonder if we can do away with the type itself to avoid the obvious trouble that comes with passing it through a struct.

**Comment by rdeodhar** [comment]
Design document attached.
[Array_Kernel_Parameters.docx](https://github.com/intel/llvm/files/4391265/Array_Kernel_Parameters.docx)

**Comment by AlexeySachkov** [comment]
Seems to be moved to intel/llvm#1423

**Pull request closed without merging**